aptos-move/move-examples/dispatching/sources/storage.move

```diff
@@ -0,0 +1,92 @@
+/// The storage module stores all the state associated with the dispatch service.
+module dispatching::storage {
+    use std::option;
+    use std::string;
+
+    use aptos_std::table::{Self, Table};
+    use aptos_std::type_info::{Self, TypeInfo};
+
+    use aptos_framework::dispatchable_fungible_asset;
+    use aptos_framework::function_info::FunctionInfo;
+    use aptos_framework::fungible_asset::{Self, Metadata};
+    use aptos_framework::object::{Self, ExtendRef, Object};
+
+    friend dispatching::engine;
+
+    struct Dispatcher has key {
+        /// Tracks the input type to the dispatch handler.
+        dispatcher: Table<TypeInfo, Object<Metadata>>,
+        /// Used to store temporary data for dispatching.
+        obj_ref: ExtendRef,
+    }
+
+    /// Store the data to dispatch here.
+    struct Storage<phantom T> has drop, key {
+        data: vector<u8>,
+    }
+
+    /// Register a `T` to callback. Providing an instance of `T` guarantees that only the
+    /// originating module can call `register` for that type.
+    public fun register<T: drop>(callback: FunctionInfo, _proof: T) acquires Dispatcher {
+        let typename = type_info::type_name<T>();
```

Comment:
@davidiw apologies for chiming in. Is it important to validate that T and the callback are from the same module and/or address?
e.g. the equivalent to
```move
let info = type_info::type_of<T>();
assert!(callback.module_address == type_info::account_address(info) && callback.module_name == type_info::module_address(info), 0)
```
If the callback and proof were from different addresses, it appears this could break the safety guarantee that only the callback can execute `retrieve()`, but I'm not certain.

If some hypothetical module had a public function that returned a `FunctionInfo` for some other purpose, couldn't another module successfully execute `register()` by providing any arbitrary `T`?